### ISSUE TITLE
bug: InvalidAccessKeyId while using terragrunt with s3 as remote state

### ISSUE THREAD

**Issue by nmofonseca** [localstack/localstack#7127]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```bash
INFO[0000] The stack at /dev/iac/aws/terraform/eu-west-2/devops/inputs_test will be processed in the following order for command apply:
Group 1
- Module /home/nuno4793/dev/iac/aws/terraform/eu-west-2/devops/inputs_test
 
Are you sure you want to run 'terragrunt apply' in each folder of the stack described above? (y/n) y
Remote state S3 bucket remote-state does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) y
ERRO[0005] Create S3 bucket with retry remote-state returned an error: InvalidAccessKeyId: The AWS Access Key Id you provided does not exist in our records.
        status code: 403, request id: Q7CY0RY8PXTH8DYC, host id: sENROE+W+H/NGS/nDbFBPJs34JGbhIHALcNoCoP/Rr2VOU3n273oM2XE/mXHP7JW56+QJBQKMhQ=. Sleeping for 10s and will try again.  prefix=dev/iac/aws/terraform/eu-west-2/devops/inputs_test] 

```

This has been mentioned on another issue that however has been closed, actually doesn't have a solution

https://github.com/localstack/localstack/issues/2540

### Expected Behavior

To create Remote state bucket and DynamoDb table for lock state.

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

compose file:
```yaml
version: "3.8"

services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack
    ports:
      - "127.0.0.1:4566:4566"            # LocalStack Gateway
      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
      - "127.0.0.1:53:53"                # DNS config (only required for Pro)
      - "127.0.0.1:53:53/udp"            # DNS config (only required for Pro)
      - "127.0.0.1:443:443"              # LocalStack HTTPS Gateway (only required for Pro)
    environment:
      - DEBUG=${DEBUG-}
      - PERSISTENCE=${PERSISTENCE-}
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
      - LOCALSTACK_API_KEY=${LOCALSTACK_API_KEY-}  # only required for Pro
      - DOCKER_HOST=unix:///var/run/docker.sock
      - HTTPS_PROXY=${HTTPS_PROXY}
      - HTTP_PROXY=${HTTP_PROXY}
      - https_proxy=${HTTPS_PROXY}
      - http_proxy=${HTTP_PROXY}
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-./volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
 ```

Start localstack:

docker compose -f localstack.yaml up -d

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

terragrunt.hcl:

```hcl
remote_state {
  backend = "s3"
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite"
  } 
  config = {
    bucket         = "${local.prefix}-${local.environment}-remote-state"
    key            = "${local.aws_region}/${basename(path_relative_to_include())}/terraform.tfstate"
    region         = "eu-west-2"
    encrypt        = true
    dynamodb_table = "${local.prefix}-remote-state-lock"
  }
}

generate "provider.tf" {
  path = "provider.tf"
  if_exists = "overwrite_terragrunt"
  contents = <<EOF
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~>${local.provider_vars.locals.aws_version}"
    }
    %{ if local.provider_vars.locals.include_postgres }
    postgresql = {
      source = "cyrilgdn/postgresql"
      version = "1.17.1"
    }
    %{ endif }
  }
}

# Configure the AWS Provider
provider "aws" {
  access_key = "test"
  secret_key = "test"
  region  = "${local.aws_region}"
  skip_credentials_validation = true
  skip_metadata_api_check     = true
  skip_requesting_account_id  = true

  endpoints {

    acm            = "http://localhost:4566"
    eks            = "http://localhost:4566"  
    apigateway     = "http://localhost:4566"
    cloudformation = "http://localhost:4566"
    cloudwatch     = "http://localhost:4566"
    dynamodb       = "http://localhost:4566"
    es             = "http://localhost:4566"
    firehose       = "http://localhost:4566"
    iam            = "http://localhost:4566"
    kinesis        = "http://localhost:4566"
    lambda         = "http://localhost:4566"
    route53        = "http://localhost:4566"
    redshift       = "http://localhost:4566"
    s3             = "http://localhost:4566"
    secretsmanager = "http://localhost:4566"
    ses            = "http://localhost:4566"
    sns            = "http://localhost:4566"
    sqs            = "http://localhost:4566"
    ssm            = "http://localhost:4566"
    stepfunctions  = "http://localhost:4566"
    sts            = "http://localhost:4566"
    rds            = "http://localhost:4566"
    ec2            = "http://localhost:4566"
  }  
}
EOF
}
```

### Environment

```markdown
- OS:"Ubuntu 20.04.5 LTS"
- LocalStack: latest ("1.2.1.dev")
```

**Comment by steffyP** [comment]
Hi @nmofonseca,

unfortunately, I am not familiar with terragrunt. I tried to run an example with the config you provided and get the same result.
However, I also do not see any request incoming to LocalStack. Could you please double-check if there might be an configuration error in the with the endpoints? 🙏

**Comment by nmofonseca** [comment]
> Hi @nmofonseca,
> 
> unfortunately, I am not familiar with terragrunt. I tried to run an example with the config you provided and get the same result. However, I also do not see any request incoming to LocalStack. Could you please double-check if there might be an configuration error in the with the endpoints? 🙏

I can't see any calls on localstack either, not really sure of what is happening on the background, but my expectation would be to be able  to use remote state with terragrunt since it's just a wrapper for terraform. 

Can you explain a little bit more on what I should check in terms of endpoints?

**Comment by steffyP** [comment]
Hi @nmofonseca,

in particular I meant if the configuration for `endpoints` for the aws provider in your `terragrunt.hcl` is taken into consideration. 
What I could find so far is that the s3-endpoint config defined in your example does not work at least.

I tried setting the `endpoint` explicitly for the s3 config, like described in the [terragrunt docs](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#remote_state):
```
remote_state {
  backend = "s3"
  config = {
    endpoint       = "http://localhost.localstack.cloud:4566"
    [ ... more config here ... ]
  }
}
```
Using this endpoint configuration, I can see incoming requests to LocalStack. However, there seems to be another bug in terragrunt, that was [discussed here](https://github.com/gruntwork-io/terragrunt/issues/1191). The AWS Caller Identity is used for various calls, and it does not consider the configuration for the s3 backend.
I tried to exclude some of the calls, that will use the different caller identity, eg.:
```
remote_state {
  backend = "s3"
  config = {
   [.. more config here...]
    encrypt        = false
    skip_bucket_root_access = true
    skip_bucket_ssencryption = true
    skip_bucket_enforced_tls = true
    skip_bucket_public_access_blocking = true
  }
}

```
However, when I run this I still get exception `UnrecognizedClientException: The security token included in the request is invalid.` Which still seems to send the request to the real AWS, as there are no related logs in LocalStack. 

The bucket itself gets created on LocalStack though.

I am sorry to say that this behavior seems to be very specific to `terragrunt`, which is currently also [not in the list of officially supported integrations](https://docs.localstack.cloud/integrations/).
